witness/golang/cmd/witness/internal/witness/witness_test.go

```diff
@@ -0,0 +1,287 @@
+	ns, err := note.NewSigner(skey)
+	if err != nil {
+		return nil, fmt.Errorf("signChkpt: couldn't create signer")
+	}
+	signed := make([][]byte, len(chkpts))
+	for i, c := range chkpts {
+		s, err := note.Sign(&note.Note{Text: c}, ns)
+		if err != nil {
+			return nil, fmt.Errorf("signChkpt: couldn't sign note")
+		}
+		signed[i] = s
+	}
+	return signed, nil
+}
+
+func newOpts(d *sql.DB, logID string, logPK string, wSK string) (Opts, error) {
+	ns, err := note.NewSigner(wSK)
+	if err != nil {
+		return Opts{}, fmt.Errorf("setOpts: couldn't create a note signer")
+	}
+	h := hasher.DefaultHasher
+	logV, err := note.NewVerifier(logPK)
+	if err != nil {
+		return Opts{}, fmt.Errorf("setOpts: couldn't create a log verifier")
+	}
+	sigVs := []note.Verifier{logV}
+	log := LogInfo{
+		SigVs: sigVs,
+		LogV:  logverifier.New(h),
+	}
+	m := map[string]LogInfo{logID: log}
+	opts := Opts{
+		DB:        d,
+		Signer:    ns,
+		KnownLogs: m,
+	}
+	return opts, nil
+}
+
+func newOptsAndKeys(d *sql.DB, logID string, logPK string) (Opts, error) {
+	wSK, _, err := note.GenerateKey(rand.Reader, "witness")
+	if err != nil {
+		return Opts{}, fmt.Errorf("couldn't generate witness keys")
+	}
+	return newOpts(d, logID, logPK, wSK)
+}
+
+func newWitness(t *testing.T, d *sql.DB, logID string, logPK string) *Witness {
+	opts, err := newOptsAndKeys(d, logID, logPK)
+	if err != nil {
+		t.Fatalf("couldn't create witness opt struct: %v", err)
+	}
+	w, err := New(opts)
+	if err != nil {
+		t.Fatalf("couldn't create witness: %v", err)
+	}
+	return w
+}
+
+// dh is taken from https://github.com/google/trillian/blob/master/merkle/logverifier/log_verifier_test.go.
+func dh(h string, expLen int) []byte {
+	r, err := hex.DecodeString(h)
+	if err != nil {
+		panic(err)
+	}
+	if got := len(r); got != expLen {
+		panic(fmt.Sprintf("decode %q: len=%d, want %d", h, got, expLen))
+	}
+	return r
+}
+
+func mustCreateDB(t *testing.T) (*sql.DB, func() error) {
+	t.Helper()
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Fatalf("failed to open temporary in-memory DB: %v", err)
+	}
+	return db, db.Close
+}
+
+func TestGetChkpt(t *testing.T) {
+	for _, test := range []struct {
+		desc    string
+		setID   string
+		queryID string
+		c       Chkpt
+		want    string
+	}{
+		{
+			desc:    "happy path",
+			setID:   "testlog",
+			queryID: "testlog",
+			c:       initChkpt,
+			want:    "there",
+		}, {
+			desc:    "other log",
+			setID:   "testlog",
+			queryID: "otherlog",
+			c:       initChkpt,
+			want:    "not there",
+		}, {
+			desc:    "nothing there",
+			setID:   "testlog",
+			queryID: "testlog",
+			c:       initChkpt,
+			want:    "not there",
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			d, closeFn := mustCreateDB(t)
+			defer closeFn()
+			ctx := context.Background()
+			// Set up log keys and sign checkpoint.
+			logSK, logPK, err := note.GenerateKey(rand.Reader, test.setID)
+			if err != nil {
+				t.Errorf("couldn't generate log keys: %v", err)
```

Comment:
super nit: all these `t.Error/Fatal/Info` lines are logging lines, so they _can_ have a capital letter starting the sentence :)

Reply:
I'm not going to change it here but this is good to know for the future :).